fvm/state/state.go

```diff
@@ -0,0 +1,162 @@
+package state
+
+import (
+	"github.com/onflow/flow-go/model/flow"
+)
+
+const (
+	defaultMaxKeySize         = 16_000     // ~16KB
+	defaultMaxValueSize       = 32_000_000 // ~32MB
+	defaultMaxInteractionSize = 64_000_000 // ~64MB
+)
+
+type StateOption func(st *State) *State
+type State struct {
+	ledger                Ledger
+	draft                 map[string]payload
+	interactionUsed       uint64
+	maxKeySizeAllowed     uint64
+	maxValueSizeAllowed   uint64
+	maxInteractionAllowed uint64
+}
+
+func defaultState(ledger Ledger) *State {
+	return &State{
+		ledger:                ledger,
+		interactionUsed:       uint64(0),
+		draft:                 make(map[string]payload, 0),
+		maxKeySizeAllowed:     defaultMaxKeySize,
+		maxValueSizeAllowed:   defaultMaxValueSize,
+		maxInteractionAllowed: defaultMaxInteractionSize,
+	}
+}
+
+// NewState constucts a new state
+func NewState(ledger Ledger, opts ...StateOption) *State {
+	ctx := defaultState(ledger)
+	for _, applyOption := range opts {
+		ctx = applyOption(ctx)
+	}
+	return ctx
+}
+
+// WithMaxKeySizeAllowed sets limit on max key size
+func WithMaxKeySizeAllowed(limit uint64) func(st *State) *State {
+	return func(st *State) *State {
+		st.maxKeySizeAllowed = limit
+		return st
+	}
+}
+
+// WithMaxValueSizeAllowed sets limit on max value size
+func WithMaxValueSizeAllowed(limit uint64) func(st *State) *State {
+	return func(st *State) *State {
+		st.maxValueSizeAllowed = limit
+		return st
+	}
+}
+
+// WithMaxInteractionAllowed sets limit on total byte interaction with ledger
+func WithMaxInteractionAllowed(limit uint64) func(st *State) *State {
```

Comment:
```suggestion
func WithMaxInteractionSizeAllowed(limit uint64) func(st *State) *State {
```

I think this should also include "size" to be more descriptive and consistent with the options above

Reply:
fixed at 5127951